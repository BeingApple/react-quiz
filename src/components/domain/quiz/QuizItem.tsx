import { Quiz } from "@/types/quiz-types"
import { Button, Typography, Card, CardContent, Divider, Box, CardActionArea, Grid } from "@mui/material"
import {decode} from 'html-entities';
import QuizPaper from "../../ui-components/QuizPaper";
import { addIndex, onCorrect, onWrong } from "@/store/reducers/answers";
import { useState, useCallback, useEffect } from "react";
import { useDispatch } from "react-redux";

type Props = {
  item?: Quiz
}

const QuizItem = ({item}: Props) => {
  const dispatch = useDispatch()

  const [isCorrect, setCorrect] = useState<boolean>();
  const [isAnswered, setAnswered] = useState(false);

  const onClickAnswer = useCallback((answer: string) => {
    const isCorrect = item!.correct_answer === answer
    const answerData = {select_answer: answer, ...item}

    setCorrect(isCorrect)
    setAnswered(true)
    
    dispatch(isCorrect ? onCorrect(answerData) : onWrong(answerData))
  }, [dispatch, item])

  const onNext = useCallback(() => {
    dispatch(addIndex())
  }, [dispatch])

  useEffect(() => {
    return () => {
      setCorrect(undefined)
      setAnswered(false)
    }
  }, [item])

  return (
    <QuizPaper>
        <Box sx={{p: 2}}>
          <Typography sx={{ fontSize: 14 }} color="text.secondary" gutterBottom>{decode(item?.category)}</Typography>
          <Typography variant="h5" component="div">{decode(item?.question)}</Typography>
        </Box>
        <Divider />
        <Box sx={{p: 2}}>
          {isAnswered ? (
            <Box sx={{textAlign: 'center'}}>
              <Typography variant="h5" component="div" color={isCorrect? "green" : "red"}>{isCorrect ? '정답' : '오답'}</Typography>
              <Button variant="outlined" size="large" onClick={onNext} sx={{m: 2}}>다음 문항</Button>
            </Box>
          ) : (
            <Grid container spacing={2} sx={{justifyContent: 'center'}}>
              {item?.answers?.map((answer, i) => (
                <Grid item key={i} xs="auto">
                  <Card sx={{p: 2, width: 200, height: "100%"}}>
                    <CardActionArea onClick={() => onClickAnswer(answer)}>
                      <CardContent>
                        <Typography gutterBottom variant="h5" component="div">{i+1}</Typography>
                        <Typography variant="body2" color="text.secondary">{decode(answer)}</Typography>
                      </CardContent>
                    </CardActionArea>
                  </Card>
                </Grid>
              ))}
            </Grid>
          )}
        </Box>
    </QuizPaper>
  )
}

export default QuizItem