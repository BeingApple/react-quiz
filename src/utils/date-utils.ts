import moment from "moment"
import { Moment, unitOfTime } from "moment"

type Unit = {
  unit: unitOfTime.Base
  string: string
}

const units: Array<Unit> = [
  {unit: "years", string: "년"},
  {unit: "months", string: "개월"}, 
  {unit: "days", string: "일"}, 
  {unit: "hours", string: "시간"}, 
  {unit: "minutes", string: "분"},
  {unit: "seconds", string: "초"},
]

export const durationToString = (duration: moment.Duration) => {
  let string = ""

  for (const unit of units) {
    const unitTime = duration.get(unit.unit)
    if (unitTime > 0) {
      string += unitTime + unit.string + " "
    }
  }

  return string.trim();
}

export const dateToString = (date?: Moment) => {
  date = (typeof date === 'string' || date instanceof String) ? moment(date) : date

  return date?.format("yyyy-MM-DD HH:mm:ss") ?? ""
}