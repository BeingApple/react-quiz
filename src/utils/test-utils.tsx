import React, { PropsWithChildren } from 'react'
import { render } from '@testing-library/react'
import type { RenderOptions } from '@testing-library/react'
import { Provider as ReduxProvider } from 'react-redux'

import { setupStore, type AppStore, type RootState } from '@/store/store'
// As a basic setup, import your same slice reducers
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>
  store?: AppStore
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  const queryClient = new QueryClient()

  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return (
      <ReduxProvider store={store}>
        <QueryClientProvider client={queryClient}>
          {children}
        </QueryClientProvider>
      </ReduxProvider>
    )
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}