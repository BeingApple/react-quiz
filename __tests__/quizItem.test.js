import '@testing-library/jest-dom'
import QuizItem from '@/components/domain/quiz/QuizItem'
import { fireEvent } from '@testing-library/react'
import { renderWithProviders } from '../src/utils/test-utils'

describe('<QuizItem />', () => {
  const sampleItem = {
    type: 'multiple',
    difficulty: 'medium',
    category: 'TEST CATEGORY',
    question: 'TEST QUESTION',
    correct_answer: 'CORRENT ANSWER',
    incorrect_answers: ['INCORRECT ANSWER 1', 'INCORRECT ANSWER 2', 'INCORRECT ANSWER 3'],
    answers: ['CORRENT ANSWER', 'INCORRECT ANSWER 1', 'INCORRECT ANSWER 2', 'INCORRECT ANSWER 3'],
  }

  const setup = (props = {}) => {
    const initialProps = { item: sampleItem };
    const utils = renderWithProviders(<QuizItem {...initialProps} {...props} />)

    const { getByText, queryByText } = utils

    const item = props.item || initialProps.item

    const category = getByText(item.category)
    const question = getByText(item.question)
    const answers = item.answers.map(a => getByText(a))

    const lastAnswerText = `내가 지난 번에 고른 답 : ${item.select_answer}`
    const correctAnswerText = `정답은 ${item.correct_answer}입니다`

    return {
      ...utils,
      category,
      question,
      answers,
      lastAnswerText,
      correctAnswerText
    };
  };

  it('matches snapshot', () => {
    const { container } = setup()
    expect(container).toMatchSnapshot();
  });

  it('has category, question, answers', () => {
    const { category, question, answers } = setup()
    expect(category).toBeTruthy()
    expect(question).toBeTruthy()

    for (const answer of answers) {
      expect(answer).toBeTruthy()
    }
  });

  it('click correct answer', () => {
    const { getByText, answers } = setup()
    const correctButton = answers[0]

    fireEvent.click(correctButton)
    getByText('정답')
    getByText('다음 문항')
  });

  it('click incorrect answer', () => {
    const { getByText, answers } = setup()
    const incorrectButton = answers[1]

    fireEvent.click(incorrectButton)
    getByText('오답')
    getByText('다음 문항')
  });

  it('click correct answer when note', () => {
    const { getByText, answers, lastAnswerText, correctAnswerText } = setup({item : {select_answer: 'INCORRECT ANSWER 1', ...sampleItem}, isNote: true})
    const correctButton = answers[0]

    fireEvent.click(correctButton)
    getByText('정답')
    getByText('다음 문항')
    getByText(lastAnswerText)
    getByText(correctAnswerText)
  });

  it('click incorrect answer when note', () => {
    const { getByText, answers, lastAnswerText, correctAnswerText } = setup({item : {select_answer: 'INCORRECT ANSWER 1', ...sampleItem}, isNote: true})
    const incorrectButton = answers[1]

    fireEvent.click(incorrectButton)
    getByText('오답')
    getByText('다음 문항')
    getByText(lastAnswerText)
    getByText(correctAnswerText)
  });
});