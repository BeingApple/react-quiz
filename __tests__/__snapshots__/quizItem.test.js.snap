// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QuizItem /> matches snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1pnr7o4-MuiPaper-root"
  >
    <div
      class="MuiBox-root css-19midj6"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-16m3jxf-MuiTypography-root"
      >
        TEST CATEGORY
      </p>
      <div
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        TEST QUESTION
      </div>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
    <div
      class="MuiBox-root css-19midj6"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-12n0uxm-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-auto css-1wrgmsj-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-zd4n71-MuiPaper-root-MuiCard-root"
          >
            <button
              class="MuiButtonBase-root MuiCardActionArea-root css-1oitmrd-MuiButtonBase-root-MuiCardActionArea-root"
              tabindex="0"
              type="button"
            >
              <div
                class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
              >
                <div
                  class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
                >
                  1
                </div>
                <p
                  class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
                >
                  CORRENT ANSWER
                </p>
              </div>
              <span
                class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-auto css-1wrgmsj-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-zd4n71-MuiPaper-root-MuiCard-root"
          >
            <button
              class="MuiButtonBase-root MuiCardActionArea-root css-1oitmrd-MuiButtonBase-root-MuiCardActionArea-root"
              tabindex="0"
              type="button"
            >
              <div
                class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
              >
                <div
                  class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
                >
                  2
                </div>
                <p
                  class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
                >
                  INCORRECT ANSWER 1
                </p>
              </div>
              <span
                class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-auto css-1wrgmsj-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-zd4n71-MuiPaper-root-MuiCard-root"
          >
            <button
              class="MuiButtonBase-root MuiCardActionArea-root css-1oitmrd-MuiButtonBase-root-MuiCardActionArea-root"
              tabindex="0"
              type="button"
            >
              <div
                class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
              >
                <div
                  class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
                >
                  3
                </div>
                <p
                  class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
                >
                  INCORRECT ANSWER 2
                </p>
              </div>
              <span
                class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-auto css-1wrgmsj-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-zd4n71-MuiPaper-root-MuiCard-root"
          >
            <button
              class="MuiButtonBase-root MuiCardActionArea-root css-1oitmrd-MuiButtonBase-root-MuiCardActionArea-root"
              tabindex="0"
              type="button"
            >
              <div
                class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
              >
                <div
                  class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
                >
                  4
                </div>
                <p
                  class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
                >
                  INCORRECT ANSWER 3
                </p>
              </div>
              <span
                class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
